numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1464,8 +1430,121 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+         *
+         * Comparisons can raise errors when element-wise comparison is not
+         * possible. Some of these, though, should not be passed on.
+         * In particular, the ufuncs do not have loops for flexible dtype,
+         * so those should be treated separately.  Furthermore, for EQ and NE,
+         * we should never fail.
+         *
+         * Our ideal behaviour would be:
+         *
+         * 1. For EQ and NE:
+         *   - If self and other are scalars, return NotImplemented,
+         *     so that python can assign True of False as appropriate.
+         *   - If either is an array, return an array of False or True.
+         *
+         * 2. For LT, LE, GE, GT:
+         *   - If self or other was flexible, return NotImplemented
+         *     (as is in fact the case), so python can raise a TypeError.
+         *   - If other is not convertible to an array, pass on the error
+         *     (MHvK, 2018-06-18: not sure about this, but it's what we have).
+         *
+         * However, for backwards compatibilty, we cannot yet return arrays,
+         * so we raise warnings instead.  Furthermore, we warn on python2
+         * for LT, LE, GE, GT, since fall-back behaviour is poorly defined.
+         */
+        PyObject *exc, *val, *tb;
+        int other_is_flexible, ndim_other;
+
+        PyErr_Fetch(&exc, &val, &tb);
+        /*
+         * Determine whether other has a flexible dtype; here, inconvertible
+         * is counted as inflexible.  (This repeats work done in the ufunc,
+         * but OK to waste some time in an unlikely path.)
+         */
+        array_other = (PyArrayObject *)PyArray_FROM_O(other);
+        if (array_other) {
+            other_is_flexible = PyTypeNum_ISFLEXIBLE(
+                PyArray_DESCR(array_other)->type_num);
+            ndim_other = PyArray_NDIM(array_other);
+            Py_DECREF(array_other);
+        }
+        else {
+            PyErr_Clear(); /* we restore the original error if needed */
+            other_is_flexible = 0;
+            ndim_other = 0;
+        }
+        if (cmp_op == Py_EQ || cmp_op == Py_NE) {
+            /* note: for == and !=, a flexible self cannot get here */
+            if (other_is_flexible) {
+                /*
+                 * For scalars, returning NotImplemented is correct.
+                 * For arrays, we emit a future deprecation warning.
+                 */
+                if (ndim_other != 0 || PyArray_NDIM(self) != 0) {
+                    if (DEPRECATE_FUTUREWARNING(
+                            "elementwise comparison failed; returning scalar "
+                            "instead, but in the future will perform "
+                            "elementwise comparison") < 0) {
+                        /*
+                         * In future, we should create a correctly shaped
+                         * array of bool. For now, a placeholder error.
+                         */
```

Comment:
I think this should

* Appear before the `DEPRECATE_FUTUREWARNING` line
* say `placeholder warning`, not placeholder error.

Reply:
It is just that if the deprecation is removed, the chained exception should be replaced (I wanted to have code that would at least work, but didn't feel like creating the array now...). I clarified this.

Reply:
I'd say that's obvious, in the same way that the `return NULL` should also be replaced